### DISCUSSION TITLE
Proposal for Native `whereLike` Method in Laravel Query Builder

### DISCUSSION THREAD

**Discussion by rupeshstha** [laravel/framework#51756] [Polls]
I have been working on a feature where I need a whereLike method to perform case-insensitive searches across multiple attributes. Currently, I am using a macro to extend the query builder with this functionality. Here is my implementation:
```php
// Macro for whereLike
Builder::macro('whereLike', function ($attributes, string $searchTerm) {
    $this->where(function (Builder $query) use ($attributes, $searchTerm) {
        foreach (Arr::wrap($attributes) as $attribute) {
            $query->when(
                Str::contains($attribute, '.'),
                function (Builder $query) use ($attribute, $searchTerm) {
                    $explodedAttribute = explode('.', $attribute);
                    $relationAttribute = end($explodedAttribute);
                    array_pop($explodedAttribute);
                    $relationName = (count($explodedAttribute) > 1)
                        ? implode('.', $explodedAttribute)
                        : $explodedAttribute[0];
                    $query->orWhereHas(
                        $relationName,
                        function (Builder $query) use ($relationAttribute, $searchTerm) {
                            $query->where($relationAttribute, 'LIKE', "%{$searchTerm}%");
                        }
                    );
                },
                function (Builder $query) use ($attribute, $searchTerm) {
                    $query->orWhere($attribute, 'LIKE', "%{$searchTerm}%");
                }
            );
        }
    });
    return $this;
});
```
**Proposal**
I believe that having a native whereLike method in Laravel's query builder would be beneficial for many developers, as it simplifies the process of performing case-insensitive searches and supports searching through related models.

**Benefits**
1. Simplicity: A native method would reduce the need for custom macros, leading to cleaner codebases.
2. Consistency: It would provide a standardized way to perform these types of searches across different projects.
3. Performance: A native implementation could be optimized for better performance compared to a user-defined macro.

Usage Example
Here’s how the whereLike method can be used with the current macro:
```php
// Example usage of whereLike
$users = User::whereLike(['name', 'email', 'profile.bio'], 'john')->get();
```

**Comment by ghost** [comment]
Bear in mind that any **or** or **like %ddd%** in your query messes up the index query and slows down the query. Full text search index was invented just for what you want. Eloquent does not support it so you must raw code it.

I would not put this code in laravel as it is generating a slow db query execution for sure.

**Reply by rupeshstha** [reply]
Yes you are right! column should be indexed accordingly first but, In my cases i have indexed column and i want this type of functionality to support natively. if possible 😄

**Reply by rupeshstha** [reply]
but yeah sure full text search can do the job but need to index data in search engine.

**Reply by wizzymore** [reply]
Eloquent supports both Like and Full Text Search.

Also, in some databases like PGSQL you can index like queries.

**Comment by wizzymore** [comment]
I'd just extract this in a scope to be honest, we already can use `->where($column, 'like', "%$value%")', if you want to scope this to multiple columns as once making a reusable scope is the play here.

**Comment by andreapozza** [comment]
Probably [Laravel Scout](https://laravel.com/docs/13.x/scout) can be handy.
